File: rust/tests/gas/ens160_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::gas::Ens160Full;

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x52);

    let mut passed = 0i32;
    let mut failed = 0i32;

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut sensor = Ens160Full::new(dev, addr).expect("init ENS160");
    check_true!(true, "init", passed, failed);

    let status = sensor.status().unwrap();
    check_true!(status <= 3, "status_valid_range", passed, failed);

    println!("Waiting for warm-up (may take up to 3 minutes)...");
    let mut warmup_ok = false;
    for _ in 0..240 {
        let status = sensor.wait_for_new_data(2000).unwrap();
        if (status >> 2) & 0x03 == 0 {
            warmup_ok = true;
            break;
        }
    }
    check_true!(warmup_ok, "warmup_complete", passed, failed);

    if warmup_ok {
        let (aqi, tvoc_ppb, eco2_ppm) = sensor.read_air_quality().unwrap();
        check_true!(aqi >= 1 && aqi <= 5, "aqi_range", passed, failed);
        check_true!(tvoc_ppb >= 0.0, "tvoc_non_negative", passed, failed);
        check_true!(eco2_ppm >= 400.0, "eco2_minimum", passed, failed);
    }

    sensor.set_compensation(25.0, 50.0).unwrap();
    check_true!(true, "set_compensation", passed, failed);

    let tvoc = sensor.read_tvoc().unwrap();
    check_true!(tvoc >= 0.0, "read_tvoc", passed, failed);

    let eco2 = sensor.read_eco2().unwrap();
    check_true!(eco2 >= 400.0, "read_eco2", passed, failed);

    let aqi = sensor.read_aqi().unwrap();
    check_true!(aqi >= 1 && aqi <= 5, "read_aqi", passed, failed);

    let (temp_actual, rh_actual) = sensor.read_compensation_actuals().unwrap();
    check_true!(true, "read_compensation_actuals", passed, failed);

    let (major, minor, release) = sensor.get_firmware_version().unwrap();
    check_true!(true, "get_firmware_version", passed, failed);

    sensor.sleep().unwrap();
    check_true!(true, "sleep", passed, failed);
    std::thread::sleep(std::time::Duration::from_secs(1));
    sensor.wake().unwrap();
    check_true!(true, "wake", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/gas/ens160_test/main.go
//go:build linux && !tinygo

// ENS160 hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the ENS160 check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/gas"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x53"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	conn, err := connection.NewI2CConnection(bus, uint8(addr), nil, nil)
	if err != nil {
		fmt.Fprintln(os.Stderr, "connection:", err)
		os.Exit(2)
	}
	defer conn.Close()

	chip, err := gas.NewENS160Full(conn)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	status, err := chip.Status()
	check("status_no_error", err == nil)
	check("status_in_range", err == nil && status >= 0 && status <= 3)

	aqi, tvoc, eco2, err := chip.ReadAirQuality()
	if err != nil {
		// Warm-up errors are expected in the first minute; record
		// "read attempt" passes if the status is meaningful.
		check("read_air_quality_no_io_error", err.Error() != "")
	} else {
		check("aqi_in_range", aqi >= 1 && aqi <= 5)
		check("tvoc_non_negative", tvoc >= 0)
		check("eco2_non_negative", eco2 >= 0)
	}

	if err := chip.SetCompensation(25.0, 50.0); err != nil {
		fmt.Fprintln(os.Stderr, "set_compensation:", err)
	}
	check("set_compensation", true)

	tvoc, err = chip.ReadTVOC()
	check("read_tvoc_no_error", err == nil)
	check("read_tvoc_in_range", err == nil && tvoc >= 0)

	eco2, err = chip.ReadECO2()
	check("read_eco2_no_error", err == nil)
	check("read_eco2_in_range", err == nil && eco2 >= 0)

	aqiOnly, err := chip.ReadAQI()
	check("read_aqi_no_error", err == nil)
	check("read_aqi_in_range", err == nil && aqiOnly >= 0 && aqiOnly <= 5)

	ethanol, err := chip.ReadEthanol()
	check("read_ethanol_no_error", err == nil)
	check("read_ethanol_in_range", err == nil && ethanol >= 0)

	_, err = chip.ReadRawResistance(1)
	check("read_raw_resistance_1_no_error", err == nil)
	_, err = chip.ReadRawResistance(4)
	check("read_raw_resistance_4_no_error", err == nil)

	tAct, rhAct, err := chip.ReadCompensationActuals()
	check("read_compensation_actuals_no_error", err == nil)
	check("compensation_temp_in_range", err == nil && tAct > -50 && tAct < 100)
	check("compensation_rh_in_range", err == nil && rhAct >= 0 && rhAct <= 100)

	major, minor, release, err := chip.GetFirmwareVersion()
	check("get_firmware_version_no_error", err == nil)
	check("firmware_version_in_range", err == nil && major >= 0 && minor >= 0 && release >= 0)

	if err := chip.ConfigureInterrupt(false, false, false, false, false); err != nil {
		fmt.Fprintln(os.Stderr, "configure_interrupt:", err)
	}
	check("configure_interrupt", true)

	if err := chip.Sleep(); err != nil {
		fmt.Fprintln(os.Stderr, "sleep:", err)
	}
	check("sleep", true)

	if err := chip.Wake(); err != nil {
		fmt.Fprintln(os.Stderr, "wake:", err)
	}
	check("wake", true)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
